### PR TITLE
ENH: Use `syrk` to compute certain dot products more quickly and accurately

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/numpy/numpy/issues/6794

Related: https://github.com/numpy/numpy/pull/6930

Deals with the special cases of `a.T @ a` and `a @ a.T` using `syrk` instead of `gemm`.

### CODE REVIEW

**Comment by jakirkham** [comment]
I think something is wrong with the `lda` parameter I am supplying here. If somebody knows a bit more about how this is working here for the `gemm` case can explain it to me, maybe I can get it fixed up.

Also, I have tried to check to see if an array is a transpose view using an ad hoc mechanism I came up with. If somebody has a better way to check this, I am open to suggestions.

**Comment by jakirkham** [comment]
I tried building this with OpenBLAS locally and am not seeing this same error. Also, I am seeing an appropriate speed up here (e.g. ~2x faster).

**Comment by jakirkham** [comment]
Ok, I still can't say that this is totally correct, but it doesn't seem to be complaining about the `ld*` values. However, I am still not confident that this is correct. Suggestions?

**Comment by jakirkham** [comment]
Pinging: @njsmith @charris @argriffing

**Comment by jakirkham** [comment]
Also, this probably needed a better way of checking the arrays are mere transposes of each other or some flags to suppress complaints about the current strategy if no better version can be found.

File: numpy/core/src/multiarray/cblasfuncs.c [hunk 1]
```diff
@@ -647,7 +715,31 @@ cblas_matrixproduct(int typenum, PyArrayObject *ap1, PyArrayObject *ap2,
...
             ldb = (PyArray_DIM(ap2, 0) > 1 ? PyArray_DIM(ap2, 0) : 1);
         }
-        gemm(typenum, Order, Trans1, Trans2, L, N, M, ap1, lda, ap2, ldb, ret);
+
+        /*
+         * Use syrk if we have a case of a matrix times its transpose.
+         * Otherwise, use gemm for all other cases.
+         */
+        if (
+            (ap1 == PyArray_BASE(ap2)) &&
```
Comment (njsmith) [inline numpy/core/src/multiarray/cblasfuncs.c]: Ah, yeah, this check is definitely not right. The base pointer doesn't have any semantics.

I think the correct check would be something like: same data pointer, and shapes are transposes of each other, and one is C contiguous and the other is F contiguous (the Trans checks are probably sufficient for the last part).

**Comment by charris** [comment]
@jakirkham  Trying to figure out when one matrix is the transpose of the other is a bit tricky. Furthermore, these days we tend to design functions to operate on stacks of matrices. On account of that, I think it would be better to have explicit functions that receive a single argument and perform `A^T * A` or `A^H * A` explicitly. Since it is often best if the multiplication would be enclosed in parenthesis anyway, the function also helps with that. i.e., instead of `A.T @ B.T @ B @ A` one simply writes `fun(B @ A)`. The only problem I have is the name of the function. One would think such a common operation would have a name, but I've never heard of one. Maybe msqr (to complement the matrix square root, or mouter, since it is a form or outer product, or `ata` and `aha`.

Other possibilities are two argument functions for `inv(A) @ B @ A` or `A^T @ B @ A`, with `B` optional.  The first form turns up in coordinate transforms, the second in transforming covariances and such. Anyway, I think proposals along those lines would be an excellent topic for discussion. OTOH, it isn't clear to me that the two argument forms buy anything in the way of efficiency or clarity. Hmm..., maybe they aren't such great ideas.

File: numpy/core/src/multiarray/cblasfuncs.c (hunk 1 above)
Comment (jakirkham) [inline numpy/core/src/multiarray/cblasfuncs.c]: Sorry, not super familiar with the NumPy C-API. So, this was my first attempt doing something significant with it. I'll think about how to rework it.

**Comment by jakirkham** [comment]
@charris, regarding the first case (i.e. the one argument form) I think I still agree with you, which is why I had mentioned something similar earlier. ( https://github.com/numpy/numpy/issues/6794#issuecomment-168436451 ) That being said, it wasn't and still isn't clear to me how I was going to make traction on that problem. As the initial problem proposed by @njsmith seemed a bit simpler to get traction, I tried it here. Adding a new function or function(s) to do this may be better in the long run.

I am aware of these two argument problems (including this `A.H @ B @ A`) and have used them before. I think part of the reason that `A.T @ A` and `A.H @ A` were brought up is they can be implemented using standard BLAS routines `syrk` and `herk`. These should be faster than `gemm` and guarantee some symmetry, which I believe was the initial reason this was raised. I am not aware of any BLAS routines that do these two argument transforms in total on a general matrix. That being said, there may be value in having such a convenience function, but I think that is a different problem than the one being solved here. Also, given it will still likely use `dot` it probably can just be implemented in Python.

**Comment by njsmith** [comment]
R names those operations `crossprod(A) = A.T @ A`, `tcrossprod(A) = A @ A.T`. Not sure if those names are worth stealing or not.

I think that's an orthogonal issue though. I don't see any reason *not* to optimize `A @ A.T` if that's what someone writes? The stacked matrices thing shouldn't be an issue because that just consists of a loop around the core operation, and the core operation is what we're talking about optimizing. And I don't see any particular challenge in detecting transposes... a cheap sufficient check is just `a.data == b.data and a.shape == b.shape[::-1] and a.strides == b.strides[::-1]`. (Possibly the strides check can be relaxed further, I'd have to think about it more. But this should catch at least 99% of cases.)

**Comment by jakirkham** [comment]
> I don't see any reason not to optimize A @ A.T if that's what someone writes?

This I definitely agree with, which is why I took the time to write this code at all. Regarding the optimization stuff, I believe both of you know more than me in this area so I'll let you guys hash it out. As for the checks, I think I have fixed them, but they probably should be fused into one conditional to reduce the rechecking.

**Comment by njsmith** [comment]
Might be good to add tests for the trickier cases like, `A @ A[:-1, :].T`, `A @ A[:, :-1].T` have matching data and strides but not shape, `A[::2, :] @ `A[:A.shape[0] // 2, :].T` have matching data and shapes but not strides, etc.

**Comment by jakirkham** [comment]
Ok, sure that's a good idea. I'll try to work on the tests tomorrow. I've tried to combine the two checks a bit so that they are not repeated for each case.

**Comment by jakirkham** [comment]
@njsmith, I've added some tests that you have suggested. They have checked out locally. Travis should be done soon. If there are any more you can think of, please let me know.

**Comment by jakirkham** [comment]
Travis checks out.

**Comment by njsmith** [comment]
This is looking good to me. (@charris: do you still have any objections?)

@jakirkham: One last thing: this is a pretty awesome user-visible feature, so we should advertise it in the release notes. Could you add a short note to `doc/release/1.11.0-notes.rst`?

**Comment by njsmith** [comment]
...I guess if you're feeling inspired you could also usefully add some benchmarks to the `benchmarks/` directory for `dot(A, A.T)` and functions that use `dot(A, A.T)`, like `cov`. But I don't really know the details of how this works, and it's not necessary for merging.

**Comment by jakirkham** [comment]
> One last thing: this is a pretty awesome user-visible feature, so we should advertise it in the release notes. Could you add a short note to `doc/release/1.11.0-notes.rst`?

Absolutely, I added a note under new features. Though I could see it going under improvements. Please let me know if I need to move it or change wording.

> ...I guess if you're feeling inspired you could also usefully add some benchmarks to the `benchmarks/` directory for `dot(A, A.T)` and functions that use `dot(A, A.T)`, like `cov`. But I don't really know the details of how this works, and it's not necessary for merging.

Sure, I have no problem doing this, but I may need a pointer on how to go about doing this. In the worst, case I can copy and tweak existing benchmarks to serve our ends here.

**Comment by jakirkham** [comment]
Rebased.

**Comment by charris** [comment]
@njsmith The one question I had kicking around in my mind was the proper definition of `transpose` in this situation when the array is of dimension > 2. That was actually a question raised at numpy's enception, with the two possibilities being: 1) reverse the last two indices (stacked), 2) reverse all indices. The second was decided on, but since `@` works on stacked arrays, we might also want to detect the case when just the last two indices are reversed. That doesn't apply to this PR, however, currently dot is only using cblas for one and two dimensional arrays. So I have no objections as long as there are no bugs ;)

**Comment by jakirkham** [comment]
Are you thinking of things like `tensordot`?

**Comment by charris** [comment]
@jakirkham I wasn't, but feel free ;) Einsum is another function we might want to look at some day.

**Comment by njsmith** [comment]
@charris: I think that all works out automagically... the way this PR works, if you somehow, by whatever means, end up in `cblas_matrixproduct` with two arrays (and these arrays might be slices of some larger array that's been passed to a gufunc, or whatever), and those two arrays fulfill the conditions needed for `syrk` to be applied, then it will be.

The only question is that IIRC we might have multiple ways of doing dot product like operations of which only some go through `cblas_matrixproduct`, so there might be some refactoring needed to make this fully universal. If `dot` learns to handle stacked matrices by doing a loop around `cblas_matrixproduct`, then the `syrk` changes automatically apply.

**Comment by jakirkham** [comment]
As far as I can tell `tensordot` falls into this first class that @njsmith has described. It will benefit from `syrk` if the transposing and reshaping it does is not too aggressive (i.e. ends up copying) and meets our constraints. We could look into more tests for `tensordot` as I only see one and I think is intended to test `einsum`.

Regarding `einsum`, I don't see any mention of using a BLAS in the source code nor does it seem to include any relevant headers for doing so. Also, I don't see any mention of `cblas_matrixproduct`. However, I am not sure how the templating here is working so maybe something gets inserted? I doubt this is the case, but I raise it anyways.

**Comment by rgommers** [comment]
Nice improvement! It'd be useful to add an ``asv`` benchmark for this case I think.

> R names those operations ``crossprod(A) = A.T @ A``, ``tcrossprod(A) = A @ A.T``. Not sure if those names are worth stealing or not.

Don't think so - much less readable, so unless you can get an additional speedup why would you introduce those names?

> Absolutely, I added a note under new features. Though I could see it going under improvements. Please let me know if I need to move it or change wording.

Not super important, but I'd call this an improvement.

**Comment by njsmith** [comment]
Yeah, currently einsum never uses BLAS; it has a pretty naive high-level
evaluation strategy. There's a lot of room to improve it by adding a
top-level execution planner to select an efficient evaluation order and to
identify sub-problems that can exploit BLAS primitives, but that'd be a
somewhat larger project...

On Tue, Jan 5, 2016 at 11:15 PM, Ralf Gommers <notifications@github.com>
wrote:

> Nice improvement! It'd be useful to add an asv benchmark for this case I
> think.
>
> R names those operations crossprod(A) = A.T @ A, tcrossprod(A) = A @ A.T.
> Not sure if those names are worth stealing or not.
>
> Don't think so - much less readable, so unless you can get an additional
> speedup why would you introduce those names?
>
> Absolutely, I added a note under new features. Though I could see it going
> under improvements. Please let me know if I need to move it or change
> wording.
>
> Not super important, but I'd call this an improvement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6932#issuecomment-169255048>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

**Comment by jakirkham** [comment]
> Nice improvement! It'd be useful to add an `asv` benchmark for this case I think.

Thanks, @rgommers. I might need some pointers on how to go about adding something to the benchmarking suite.

> > R names those operations `crossprod(A) = A.T @ A`, `tcrossprod(A) = A @ A.T`. Not sure if those names are worth stealing or not.
> Don't think so - much less readable, so unless you can get an additional speedup why would you introduce those names?
@njsmith is referring to a longer discussion in ( https://github.com/numpy/numpy/issues/6794 ); where, the merits of optimizing these expressions were discussed `A.T @ A`, `A @ A.T`, `A.H @ A`, `A @ A.H`. The first two can be optimized with `syrk` as here. The last two can be optimized with `herk`. As the last two aren't views (unless the matrices are real and then are just the first two cases), there wasn't an easy way to check for the last two cases here. So, there were some discussions about adding a function or other modifications so as to handle all the cases.

> Not super important, but I'd call this an improvement.

Thanks for the feedback. This is now done.

**Comment by jakirkham** [comment]
@njsmith, ah, ok, I wasn't sure initially if there was concern about having bad behavior in `einsum` or if we were seeking to optimize `einsum` and were concerned about the special case being added to `dot`.

**Comment by jakirkham** [comment]
Here are some timings with ATLAS.

    In [1]: import numpy
    
    In [2]: import numpy as np
    
    In [3]: a = np.ones((1000,1001))
    
    In [4]: %timeit np.dot(a.T, a)
    10 loops, best of 3: 56.9 ms per loop
    
    In [5]: %timeit np.dot(a.T.copy(), a)
    10 loops, best of 3: 96.1 ms per loop
    
    In [6]: %timeit np.dot(a.T, a)
    10 loops, best of 3: 54.8 ms per loop
    
    In [7]: %timeit np.dot(a.T.copy(), a)
    10 loops, best of 3: 95 ms per loop

**Comment by pv** [comment]
Note that that benchmark includes the time for `.copy()`, so it's not fully apples-to-apples.

**Comment by jakirkham** [comment]
Sure. Though, its not a huge difference, but I will update it.

**Comment by jakirkham** [comment]
Here is the improved ATLAS comparison per @pv's point.

    In [1]: import numpy
    
    In [2]: import numpy as np
    
    In [3]: a = np.ones((1000,1001))
    
    In [4]: at = a.T
    
    In [5]: atc = a.T.copy()
    
    In [6]: %timeit np.dot(at, a)
    10 loops, best of 3: 54.7 ms per loop
    
    In [7]: %timeit np.dot(atc, a)
    10 loops, best of 3: 89.5 ms per loop
    
    In [8]: %timeit np.dot(a, at)
    10 loops, best of 3: 51.6 ms per loop
    
    In [9]: %timeit np.dot(a, atc)
    10 loops, best of 3: 86.5 ms per loop

**Comment by jakirkham** [comment]
Here are some timings with OpenBLAS 0.2.15, which I compiled by hand (details can be provided):

In [1]: import numpy

In [2]: import numpy as np

In [3]: a = np.ones((1000,1001))

In [4]: at = a.T

In [5]: atc = a.T.copy()

In [6]: %timeit np.dot(at, a)
10 loops, best of 3: 32.2 ms per loop

In [7]: %timeit np.dot(atc, a)
10 loops, best of 3: 51.1 ms per loop

In [8]: %timeit np.dot(a, at)
10 loops, best of 3: 31.9 ms per loop

In [9]: %timeit np.dot(a, atc)
10 loops, best of 3: 50.9 ms per loop

**Comment by jakirkham** [comment]
Of interest to @njsmith and @rgommers, I have added something benchmarks to suite. The arrays used here are kind of small so the improvement is not as noticeable as with bigger arrays. Please let me know your thoughts.

**Comment by jakirkham** [comment]
Updated the benchmarks so they would be sorted in the output next to each other to make it easy to compare.

**Comment by jakirkham** [comment]
Please let me know if there is anything else needed. Otherwise, I am assuming this is complete.

Also, I have tried to rebase against master a few times, but there seem to a bunch of updates going in. Instead of clogging up the build queue with rebases of this PR, I am going to let it sit. If you would like to merge and want me to rebase beforehand, feel free to give me a nudge and I will do so.

**Comment by njsmith** [comment]
FYI as a general rule, there's no need to rebase at all unless master has changed in a way that breaks your patch, or else you want to rewrite history for some other reason (e.g. fix commit messages or squash commits together).

Anyway, I think this is ready to go -- thanks @jakirkham!

**Comment by jakirkham** [comment]
Ah ok, I think in previous PRs that I have done to numpy I had been encouraged to rebase on master if I was out of date. I was assuming this was to keep the git graph from being a giant indecipherable mess. If that is not necessary, I will avoid doing rebases strictly to move commits onto master unless explicitly asked.

Cool, thanks for your help in guiding me in this effort @njsmith. Also, thank you everyone for your suggestions and feedback. If anything else comes up with regards to this, feel free to to let me know.

**Comment by rgommers** [comment]
Benchmarks show the 2x improvement now (for ``a.T`` and not for ``a.T.copy()`` as expected): http://pv.github.io/numpy-bench/#bench_linalg.Eindot.time_dot_trans_a_at

**Comment by jakirkham** [comment]
Oh wow, that's super cool. I didn't realize you had an awesome benchmark viewer for NumPy. Also, it is good to see it checks out. Thanks for sharing @rgommers.

**Comment by rgommers** [comment]
Yes we should advertise that better. I guess @pv didn't advertise because it's on his personal site now. But it's pretty cool indeed.

**Comment by jakirkham** [comment]
For those who plan to use this on Mac with OpenBLAS something to be aware of ( https://github.com/xianyi/OpenBLAS/issues/730 ). This does not affect Linux.

**Comment by jakirkham** [comment]
Turns out this actually optimizes `numpy.matmul`, as well. However, we didn't notice that at the time. A benchmark for `numpy.matmul` was added in this PR ( https://github.com/numpy/numpy/pull/7034 ).

**Pull request merged**